test/sqllogictest/transform/threshold_elision.slt

```diff
@@ -479,29 +479,27 @@ WITH MUTUALLY RECURSIVE
 SELECT * FROM c1
 ----
 Explained Query:
-  Return // { non_negative: true }
-    Get l0 // { non_negative: true }
+  Return // { non_negative: false }
+    Get l0 // { non_negative: false }
   With Mutually Recursive
     cte l1 =
       Union // { non_negative: false }
-        Get l0 // { non_negative: true }
+        Get l0 // { non_negative: false }
         Negate // { non_negative: false }
-          Filter (#0 > 2) // { non_negative: true }
-            Get l0 // { non_negative: true }
+          Filter (#0 > 2) // { non_negative: false }
+            Get l0 // { non_negative: false }
     cte l0 =
-      Distinct project=[#0, #1] // { non_negative: true }
-        Union // { non_negative: true }
+      Distinct project=[#0, #1] // { non_negative: false }
+        Union // { non_negative: false }
           Project (#0, #4) // { non_negative: true }
             Map ((#1 || "_init")) // { non_negative: true }
               ReadStorage materialize.public.people // { non_negative: true }
-          Threshold // { non_negative: true }
```

Comment:
We were able to elide this `Threshold` now, for some reason we could not previously. Now that it is elided, `Union` fusion occurs, and because the union has three terms now the pre-existing logic for reasoning about unions containing negates fails (it only worked for a two-input union).

Reply:
So, in theory, this might create a regression at customers. How hard would it be to extend the logic to n-ary Unions? Alternatively, could we add a feature flag?

Reply:
If both of the above are hard, then we could maybe just say that the chances of this introducing a big regression at a customer is quite low, so we can just go ahead and merge it as it is.

Reply:
No regression that I'm aware of, other than folks who rely on `EXPLAIN WITH (non_negative)`.

This PR does two things:
1. Modifies how non-negativity is determined, strictly improving it to include cases locked up in `threshold_elision.rs`.
2. Modifies how thresholds are elided, strictly improving the number we can elide.

What's happening here is that we have shifted both the test and the plan, improving both of them. There is no regression in the technical sense. 
```
old_test(old_plan) => non_negative = true
new_test(new_plan) => non_negative = false
```
This looks like a potential regression, but the old test on the new plan would also be `false`. And the new test on the old plan would still be `true`.

Reply:
> How hard would it be to extend the logic to n-ary Unions?

It's not, but this is one of the trade-offs: it would be quadratic in the the arguments to union, and each of the evaluations of the `is_superset_of` function are also potentially quadratic (based on their equality tests).

Reply:
We could relax the work to rely on the sort of provenance information introduced in https://github.com/MaterializeInc/materialize/pull/27932, but that has the potential to regress from the bespoke code here.

Reply:
> There is no regression in the technical sense.

No regression in this test, but I mean, one can imagine a situation where there is another `Threshold` on top of that node whose non-negative analysis result shifted from true to false, in which case we might no longer be able to elide that Threshold.

But this probably has a low chance, so I would be ok with risking this.

Reply:
> > How hard would it be to extend the logic to n-ary Unions?

> It's not, but this is one of the trade-offs: it would be quadratic in the the arguments to union, and each of the evaluations of the `is_superset_of` function are also potentially quadratic (based on their equality tests).

Oh, I see. Ok, then let's consider this out of scope for this PR. And at some later time we might do it, but with a limit of at most, say, 4 inputs.

Reply:
Btw, I may have misunderstood. This is not evidence of a regression, but you are right that it highlights the potential of a regression due to `Threshold` elision. This is the same potential as "better `Union` fusion" though: this transform has always had the property that more union fusion may result in less `Threshold` elision, and .. indeed it is awkward that "more `Threshold` elision" can result in "less subsequent `Threshold` elision".

I do think we should fix it, but I don't think it is any more a regression than any other change to the optimizer that might result in more `Union` fusion.

Reply:
I'd love to delete all of these, btw! Anything that relies on a certain shape of `Join` or `Union` is .. displeasing to me. I'll take a swing at implementing the robustness next, but it's the same sort of thing in a lot of transforms, our unique key determination, all over the place.